File: C++/quasi.cpp
#include <iostream>
#include <string>

using namespace std;

string LCQS[101][101];
string S, T;
string answer = "";

void calLongest(int m, int n) {
    for (int i=m; i<m+2; i++) {
        for (int j=n; j<n+2; j++) {
            if(i>=S.length() || j>=T.length()) continue;
            else if (m == 0 || n == 0) {
                if (LCQS[i][j] == "") LCQS[i][j] = S[m];
            }
            else {
                if (LCQS[i][j].length() < (LCQS[m-1][n-1]+S[m]).length()) {
                    LCQS[i][j] = LCQS[m-1][n-1]+S[m];
                }
                else if (LCQS[i][j].length() == (LCQS[m-1][n-1]+S[m]).length()) {
                    LCQS[i][j] = min(LCQS[i][j], LCQS[m-1][n-1]+S[m]);
                }
            }
            if(LCQS[i][j].length() > answer.length()) answer = LCQS[i][j];
            else if(LCQS[i][j].length() == answer.length()) answer = min(answer, LCQS[i][j]);
        }
    }
}

int main() {
    cin >> S >> T;

    for (int i=0; i<S.length(); i++) {
        for (int j=0; j<T.length(); j++) {
            if (S[i] == T[j]) calLongest(i, j);
        }
    }

    cout << answer << endl;

    return 0;
}

File: python/조고리즘/quasi.py
def solve(m, n):
    global LCQS, S, T, answer
    for i in range(m, m+2):
        for j in range(n, n+2):
            if i >= len(S) or j >= len(T):
                continue
            elif m == 0 or n == 0:
                if LCQS[i][j] == "":
                    LCQS[i][j] = S[m]
            else:
                if len(LCQS[i][j]) < len(LCQS[m-1][n-1] + S[m]):
                    LCQS[i][j] = LCQS[m-1][n-1] + S[m]
                elif len(LCQS[i][j]) == len(LCQS[m-1][n-1] + S[m]):
                    LCQS[i][j] = min(LCQS[i][j], LCQS[m-1][n-1] + S[m])
            if len(LCQS[i][j]) > len(answer):
                answer = LCQS[i][j]
            elif len(LCQS[i][j]) == len(answer):
                answer = min(answer, LCQS[i][j])


S = input()
T = input()

LCQS = [["" for _ in range(101)] for _ in range(101)]
answer = ""

for i in range(len(S)):
    for j in range(len(T)):
        if S[i] == T[j]:
            solve(i, j)

print(answer)

